src/levanter/infra/ray_tpu.py

```diff
@@ -312,9 +331,83 @@ def get_actor_name_from_actor_info(self, actor_info: SliceInfo) -> str:
+
+        if isinstance(num_slices, int):
+            self.scale_actor_pool(num_slices)
+            return
+
+        sorted_valid_sizes = sorted(num_slices)
+        max_valid_size = sorted_valid_sizes[-1]
 
+        # CHAOS MONKEY: Simulate initial scaling not getting enough slices:
+        # if len(self._actor_pool) == 0:
+        #     max_valid_size -= 1
+
+        logger.info(f"Attempting to scale to {max_valid_size} slices based on the maximum of valid sizes: {sorted_valid_sizes}")
+        try:
+            self.scale_actor_pool(max_valid_size)
+            # self.scale_actor_pool(max_valid_size)
+        except Exception as e:
+            logger.warning(f"Error when scaling to {max_valid_size} slices: {e}")
+
+        if len(self._actor_pool) in num_slices:
+            return
+
+        current_size = len(self._actor_pool)
+        feasible_sizes = [size for size in sorted_valid_sizes if size <= current_size]
+
+        if not feasible_sizes:
+            raise Exception(f"Only got {current_size} slices, which is not enough for minimum of valid sizes: {sorted_valid_sizes}")
+
+        max_feasible_size = feasible_sizes[-1]
+        logger.warning(f"Attempting to scale to {max_feasible_size} slices based on a feasible size in valid sizes: {sorted_valid_sizes}")
+        try:
+            self.scale_actor_pool(max_feasible_size)
+        except Exception as e:
+            logger.warning(f"Error when scaling to {max_feasible_size} slices: {e}")
+
+        if len(self._actor_pool) not in num_slices:
+            raise Exception(f"Could not scale to {max_feasible_size} slices")
+
+
+    def check_should_scale_up_multislice(self, num_slices: int | Sequence[int]) -> bool:
+        if isinstance(num_slices, int):
+            return False
+
+        # Don't check for rescaling if:
+        # - Not enough time has passed since the the last scale up check
+        # - Not enough time has passed since the the last actual scale up
+        current_time = time.time()
+        last_check_time = self._last_check_should_scale_up_multislice_time
+        self._last_check_should_scale_up_multislice_time = current_time
+        if last_check_time is None or current_time - last_check_time < _SCALE_UP_MULTISLICE_CHECK_INTERVAL:
+            return False
+
+        if self._last_scale_multislice_time is None or current_time - self._last_scale_multislice_time < _SCALE_UP_MULTISLICE_INTERVAL:
+            return False
+
+        # Want to scale to the next largest desired size that is bigger than the current size.
+        sorted_valid_sizes = sorted(num_slices)
+        current_size = current_size = len(self._actor_pool)
+        larger_sizes = [size for size in sorted_valid_sizes if size > current_size]
+        if not larger_sizes:
+            return False
+        next_larger_size = larger_sizes[0]
+
+        # We start N slice probe actors to probe for available slices.
+        # We shut down these actors after we're done checking.
```

Comment:
is this better than just sending them directly?

Reply:
Need to update this comment. If we are successful, we keep the actors. If we're not, we shut down the actors.